base/inc/AdePT/TrackManager.h

```diff
@@ -0,0 +1,263 @@
+  mgr->fActiveTracks = activeSlots;
+  mgr->fNextTracks   = nextSlots;
+  mgr->fBuffer       = buffer;
+  // construct the MParray objects inplace
+  adept::MParray::MakeInstanceAt(capacity, activeSlots);
+  adept::MParray::MakeInstanceAt(capacity, nextSlots);
+}
+
+template <typename Track>
+__global__ void clear_trackmanager(TrackManager<Track> *mgr)
+{
+  mgr->clear();
+}
+
+template <typename Track>
+__global__ void swap_active(TrackManager<Track> *mgr)
+{
+  mgr->swap();
+}
+
+template <typename Track>
+__global__ void defragment_buffer(TrackManager<Track> *mgr, int nactive, int where)
+{
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < nactive; i += blockDim.x * gridDim.x) {
+    const int slot_src     = (*mgr->fNextTracks)[i];
+    const int slot_dst     = (where + i) & (mgr->fCapacity - 1);
+    mgr->fBuffer[slot_dst] = mgr->fBuffer[slot_src];
+    // fActiveTracks must be cleared before starting this kernel
+    mgr->fActiveTracks->push_back(slot_dst);
+  }
+}
+
+template <typename Track>
+__global__ void adjust_indices(TrackManager<Track> *mgr, int start_new, int next_free)
+{
+  mgr->fStats.fStart = start_new;
+  mgr->fNextFree.store(next_free);
+  mgr->fNextTracks->clear();
+}
+
+} // End namespace device_impl_trackmgr
+#endif
+
+/// @brief A track manager working with a circular buffer.
+template <typename Track>
+struct TrackManager {
+
+  /// @brief Statistics of the track manager to be updated and copied to host between iterations
+  struct Stats {
+    int fStart{0};     ///< Index of first used track in the buffer
+    int fNextStart{0}; ///< Index where the tracks will be compacted
+    int fInFlight{0};  ///< Number of tracks sill in flight
+
+    int GetNused() { return fNextStart - fStart; }
+  };
+
+  Stats fStats;                       ///< Current statistics
+  int fCapacity{0};                   ///< Maximum number of elements
+  adept::Atomic_t<int> fNextFree;     ///< Index of last used index in the buffer
+  TrackManager *fInstance_d{nullptr}; ///< Device instance
+
+  adept::MParray *fActiveTracks{nullptr}; ///< Array of active (input) track slots (device pointer)
+  adept::MParray *fNextTracks{nullptr};   ///< Array of rack slots for the next iteration (device pointer)
+  Track *fBuffer{nullptr};                ///< Storage for the circular buffer of tracks (device pointer)
+
+  /// @brief Construction done on host but holding device pointers.
+  __host__ __device__ TrackManager(size_t capacity) : fCapacity(capacity)
+  {
+    // The Track type must be copy constructible
+    if (!std::is_copy_constructible<Track>::value)
+      COPCORE_EXCEPTION("TrackManager ctor: The track type must be copy constructible");
+    // The queue size must be a power of 2 (for fast access in the circular buffer)
+    bool pow_two = (capacity & (capacity - 1)) == 0;
+    if (!pow_two) COPCORE_EXCEPTION("TrackManager ctor: The buffer size has to be a power of 2");
```

Comment:
Is this really needed? Just do a modulo operation instead.

Reply:
Getting slots is called very frequently, which needs to do systematically a modulo, and modulo is expensive, so I prefer this for the time being. We can hide it behind a macro to make code more readable, maybe in the next fixup.

Reply:
> modulo is expensive

I don't agree on that, do you have numbers?

Reply:
Indeed, the difference is at per-mil level, not justifying limiting the buffer size to be a power of 2. Fixed that.